### PR TITLE
Make governance variables to be atomic

### PR DESCRIPTION
**Pull request by andybclee**
## Proposed changes
- This PR includes following PRs which were reverted before. To remove data race, several variables in the governance object have been changed to be atomic.
- Add missing atomic operation for lastGovernanceStateBlock https://github.com/ground-x/klaytn/pull/3517
- Make nodeAddress as an atomic.Value https://github.com/ground-x/klaytn/pull/3524
- Add idxCacheLock to protect idxCache https://github.com/ground-x/klaytn/pull/3526

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: governance/default.go [hunk 1]
```diff
@@ -605,13 +617,6 @@ func getGovernanceCacheKey(num uint64) common.GovernanceCacheKey {
 	return common.GovernanceCacheKey(params.GovernanceCachePrefix + "_" + v)
 }
 
-func (g *Governance) addIdxCache(num uint64) {
```
Comment (ehnuje) [inline governance/default.go]: Why did you remove this function?

Comment (andybclee) [reply governance/default.go]: @ehnuje As far as I can remember, there was an issue for adding a cache for test code after idxCacheLock was added. It looks like it could be better. So I revived addIdxCache function and make AddGovernanceCacheForTest to use `addGovernanceCache` while not accessing `addIdxCache` directly.

File: governance/default.go [hunk 2]
```diff
@@ -927,7 +938,5 @@ func AddGovernanceCacheForTest(g *Governance, num uint64, config *params.ChainCo
 	// Don't update cache if num (block number) is smaller than the biggest number of cached block number
 
 	data := getGovernanceItemsFromChainConfig(config)
-	cKey := getGovernanceCacheKey(num)
-	g.itemCache.Add(cKey, data.Items())
```
Comment (ehnuje) [inline governance/default.go]: In the last commit, 930 and 931 has been removed, is it intended?

File: governance/api.go [hunk 3]
```diff
@@ -89,7 +89,7 @@ func (api *PublicGovernanceAPI) Vote(key string, val interface{}) (string, error
 	gMode := api.governance.ChainConfig.Governance.GovernanceMode
 	gNode := api.governance.ChainConfig.Governance.GoverningNode
 
-	if GovernanceModeMap[gMode] == params.GovernanceMode_Single && gNode != api.governance.nodeAddress {
+	if GovernanceModeMap[gMode] == params.GovernanceMode_Single && gNode != api.governance.nodeAddress.Load().(common.Address) {
```
Comment (jimni1222) [inline governance/api.go]: Can 'NodeAddress()' function be used in here? 
like "gNode != api.NodeAddress()"

File: governance/default.go [hunk 4]
```diff
@@ -591,6 +596,9 @@ func (g *Governance) getGovernanceCache(num uint64) (map[string]interface{}, boo
 
 func (g *Governance) addGovernanceCache(num uint64, data GovernanceSet) {
```
Comment (jeonghonah) [inline governance/default.go]: Calling `addIdxCache()` without locking might be dangerous. How about simply embedding the `addIdxCache()` to the call site?

File: governance/api.go (hunk 3 above)
Comment (andybclee) [reply governance/api.go]: @jimni1222 Yes, but not now. In following PRs, there is a change in the function 'NodeAddress()' to use load(). I'll apply your suggestion after applying it.

File: governance/default.go (hunk 4 above)
Comment (andybclee) [reply governance/default.go]: @jeonghonah The function was called by two functions before, but after fixing it now only one function is calling it. So as you suggested, it can be merged with the caller.
Thank you for your great feedback. =)

File: governance/default.go (hunk 2 above)
Comment (andybclee) [reply governance/default.go]: @ehnuje Yes, it was intended. Because now the function calls g.addGovernanceCache(), those two removed lines is being done in the addGovernanceCache().

File: governance/default.go [hunk 5]
```diff
@@ -562,7 +564,10 @@ func (g *Governance) initializeCache() error {
 	if err != nil {
 		return ErrNotInitialized
 	}
+	g.idxCacheLock.Lock()
 	g.idxCache = indices
+	g.idxCacheLock.Unlock()
+
 	// Put governance items into the itemCache
 	for _, v := range indices {
```
Comment (laggu) [inline governance/default.go]: although it is same, would it be better use g.idxCache instead of indices?

Comment (andybclee) [reply governance/default.go]: @laggu Good point. Thanks!

Comment (ehnuje) [reply governance/default.go]: Isn't locking needed here? (at least RLock())

Comment (andybclee) [reply governance/default.go]: @ehnuje initializeCache is called only once when a node starts while calling `NewGovernance()`. So I don't think a locking is needed for `indices`.

Comment (andybclee) [reply governance/default.go]: @ehnuje I removed lock on `g.idxCacheLock` in `initializeCache()` because it is redundant as we discussed. Please take a look.

**Comment by andybclee** [comment]
@KimKyungup @jeonghonah @nogang @ehnuje @laggu @jimni1222 @aidan-kwon Please take a look. =)

File: governance/default.go [hunk 6]
```diff
@@ -563,6 +565,7 @@ func (g *Governance) initializeCache() error {
 		return ErrNotInitialized
 	}
 	g.idxCache = indices
+
```
Comment (ehnuje) [inline governance/default.go]:
```suggestion
```
